master/static/srv/ship_logs.py

```diff
@@ -0,0 +1,642 @@
+
+
+class Shipper(threading.Thread):
+    """
+    Shipper reads structured logs from logq and ships them to the determined-master.
+
+    It will send a message on doneq when it finishes.
+    """
+
+    def __init__(
+        self,
+        master_url: str,
+        token: str,
+        cert_name: str,
+        cert_file: str,
+        logq: queue.Queue,
+        doneq: queue.Queue,
+        daemon: bool,
+    ) -> None:
+        super().__init__(daemon=daemon)
+        self.logq = logq
+        self.doneq = doneq
+
+        self.headers = {"Authentication": f"Bearer {token}"}
+
+        baseurl = master_url.rstrip("/")
+        self.url = f"{baseurl}/task-logs"
+
+        self.context = None
+        if master_url.startswith("https://"):
+            # Create an SSLContext that trusts our DET_MASTER_CERT_FILE, and checks the hostname
+            # against the DET_MASTER_CERT_NAME (which may differ from the hostname in the url).
+            self.context = ssl.create_default_context()
+            if cert_file.lower() == "noverify":
+                # Don't check the master's certificate.
+                # Presently the master never sets this value for DET_MASTER_CERT_FILE, but we keep
+                # this check to be consistent with the CLI behavior.
+                self.context.verify_mode = ssl.CERT_NONE
+            elif cert_file:
+                # Explicitly trust the cert in cert_file.
+                self.context.load_verify_locations(cafile=cert_file)
+            if cert_name:
+                # Override hostname verification
+                self.context = override_verify_name(self.context, cert_name)
+
+    def run(self) -> None:
+        try:
+            self._run()
+        except Exception as e:
+            self.doneq.put(DoneMsg("shipper", error=e))
+        else:
+            self.doneq.put(DoneMsg("shipper", error=None))
+
+    def _run(self) -> None:
+        eofs = 0
+        while eofs < 2:
+            logs: List[Dict[str, Any]] = []
+            deadline = time.time() + SHIPPER_FLUSH_INTERVAL
+            # Pop logs until both collectors close, or we fill up a batch, or we hit the deadline.
+            while eofs < 2 and len(logs) < LOG_BATCH_MAX_SIZE:
+                now = time.time()
+                timeout = deadline - now
+                if timeout <= 0:
+                    # We are already passed the deadline.
+                    break
+
+                try:
+                    log = self.logq.get(timeout=timeout)
+                except queue.Empty:
+                    # We hit the timeout.
+                    break
+
+                if log is None:
+                    eofs += 1
+                    continue
+
+                logs.append(log)
+
+            data = json.dumps(logs).encode("utf8")
+
+            # Try to ship for about ten minutes.
+            backoffs = [0, 1, 5, 10, 15, 15, 15, 15, 15, 15, 15, 60, 60, 60, 60, 60, 60, 60, 60, 60]
```

Comment:
It's not weird to me that `ship` takes backoffs as a parameter (though if you wanted to make it unconfigurable, I think that would be fine, too). But I think the result where `ship` and `ship_special` have different backoffs is a surprising result.

Reply:
I am not going to pretend this is a Great Architecture Decision, rather this was more of a "where do I refactor to maximize DRY" type of decision.

In particular, I thought they should have quite different backoffs, because ship_special happens right after the task was launched, and it indicates nothing actually ran.  In my mind, that is the sort of case that is most likely going to be hit when somebody is doing heavy interactive development on a new image, new workflow, etc, and having a 10minute timeout if somebody e.g. configured an image wrong would be really unfortunate.  The normal log shipping code might start failing mid-task if, say, a master upgrade took a few minutes to run a migration.  I wanted that timeout to be pretty generous.

And so, starting with that position, having the backoffs get passed into `ship` made the most sense.